core/mapi/cache/namedPropCache.cpp

```diff
@@ -31,88 +34,182 @@ namespace cache
+					continue;
+				}
+				else if (FAILED(hRes))
+				{
+					// Other errors are unexpected
+					break;
+				}
+
 				for (const auto& name : range)
 				{
 					if (name->getMapiNameId()->lpguid != nullptr)
 					{
 						names.push_back(name);
 					}
 				}
+
+				// Go to the next batch
+				iTag += batchSize - 1;
 			}
 
+			lpMAPIFolder->Release();
 			return names;
 		}
 
+		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>>
+		GetAllNamesFromIDsFromContainer(LPMAPICONTAINER lpMAPIContainer)
+		{
+			LPMAPIFOLDER lpRootFolder = nullptr;
+			if (FAILED(GetRootFolder(lpMAPIContainer, &lpRootFolder)))
+			{
+				return {};
+			}
+
+			return GetAllNamesFromIDs(lpRootFolder);
+		}
+
+		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>> GetAllNamesFromIDsFromMdb(LPMDB lpMdb)
+		{
+			LPMAPIFOLDER lpRootFolder = nullptr;
+			if (FAILED(GetRootFolder(lpMdb, &lpRootFolder)))
+			{
+				return {};
+			}
+
+			return GetAllNamesFromIDs(lpRootFolder);
+		}
+
 		// Returns a vector of NamedPropCacheEntry for the input tags
 		// Sourced directly from MAPI
-		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>>
-		GetNamesFromIDs(_In_ LPMAPIPROP lpMAPIProp, _In_opt_ LPSPropTagArray* lppPropTags, ULONG ulFlags)
+		_Check_return_ HRESULT GetNamesFromIDs(
+			_In_ LPMAPIPROP lpMAPIProp,
+			_In_opt_ LPSPropTagArray lpPropTags,
+			ULONG ulFlags,
+			std::vector<std::shared_ptr<namedPropCacheEntry>> &names)
 		{
-			if (!lpMAPIProp) return {};
+			if (!lpMAPIProp)
+			{
+				return MAPI_E_INVALID_PARAMETER;
+			}
 
 			LPMAPINAMEID* lppPropNames = nullptr;
+			LPSPropTagArray lpPropTagsLocal = lpPropTags;
 			auto ulPropNames = ULONG{};
 
 			// Try a direct call first
-			const auto hRes =
-				WC_H_GETPROPS(lpMAPIProp->GetNamesFromIDs(lppPropTags, nullptr, ulFlags, &ulPropNames, &lppPropNames));
+			const auto hRes = WC_H_GETPROPS(
+				lpMAPIProp->GetNamesFromIDs(&lpPropTagsLocal, nullptr, ulFlags, &ulPropNames, &lppPropNames));
 
 			// If we failed and we were doing an all props lookup, try it manually instead
-			if (hRes == MAPI_E_CALL_FAILED && (!*lppPropTags))
+			if (hRes == MAPI_E_CALL_FAILED && !lpPropTags)
 			{
-				return GetAllNamesFromIDs(lpMAPIProp);
-			}
+				LPMAPICONTAINER lpContainer = nullptr;
+				HRESULT hRes2 = WC_H(lpMAPIProp->QueryInterface(IID_IMAPIContainer, reinterpret_cast<LPVOID*>(&lpContainer)));
+				if (SUCCEEDED(hRes2))
+				{
+					names = GetAllNamesFromIDsFromContainer(lpContainer);
+					lpContainer->Release();
+					return S_OK;
+				}
 
-			auto ids = std::vector<std::shared_ptr<namedPropCacheEntry>>{};
+				LPMDB lpMdb = nullptr;
+				hRes2 = WC_H(lpMAPIProp->QueryInterface(IID_IMsgStore, reinterpret_cast<LPVOID*>(&lpMdb)));
+				if (SUCCEEDED(hRes2))
+				{
+					names = GetAllNamesFromIDsFromMdb(lpMdb);
+					lpMdb->Release();
+					return S_OK;
+				}
+
+				// Can't do the get all props special route because object wasn't IID_IMAPIContainer
+				return hRes;
```

Comment:
Do we want a trace of some kind here? If so, what is an appropriate way to do that in this codebase?